### ISSUE TITLE
Call PEP 517 backend hooks via kwargs

### ISSUE THREAD

**Issue by webknjaz** [pypa/pep517#115]
I have an in-tree build backend that implements wrappers for some of the hooks in another backend. In my decorator (generic for all hooks) I use `*args, **kwargs` for the wrapper function which I then bypass to the wrapped hook call.

Now, I want to access `config_settings` in a unified fashion and use that dict somehow before proceeding to call the corresponding hooks. I noticed that the `args` tuple is populated while `kwargs` isn't. So clearly the hooks are called by `pep517` lib via positional arguments.

I propose migrating to calling all the hooks with explicit keyword arguments to accommodate use-cases like mine and this would also open up possibilities for future hook extension as the standard evolves. I think it'd be reasonable to request PEP 517 front-ends use kwargs-only calls. WDYT? Should this be something raised as a PEP improvement too?

**Comment by pradyunsg** [comment]
Could you provide a small code example of what you're doing, and how it would look after the requested change?

I'm finding the textual description hard to follow.

**Comment by pfmoore** [comment]
If I'm understanding you correctly, you're asking that *all* callers of the backend hooks must use keyword args to make the calls. That would definitely be a PEP change, but more importantly I think it's an unreasonable requirement to put on the callers. The signatures are explicit in the existing PEP and I'd expect any call that conformed to the signature to be allowed. So, for example, `build_wheel(some_dir)` should be fine, and we should not require `build_wheel(wheel_directory=some_dir)`.

You're proposing that the signature be `build_wheel(*, wheel_directory, config_settings=None, metadata_directory=None)` which is a change to the PEP.

**Comment by pradyunsg** [comment]
The other option is to pass values in this package via keyword arguments, but that's then depending on the implementation details of this package, which we really really should not do.

**Comment by pfmoore** [comment]
Agreed, a backend (or backend wrapper) that requires arguments to be passed by keyword only is not a compliant backend according to the current PEP 517 definition.

**Comment by takluyver** [comment]
We did actually think of this and make it fully explicit in the PEP :wink: :

> The hooks may be called with positional or keyword arguments, so backends implementing them should be careful to make sure that their signatures match both the order and the names of the arguments above.

You could argue that if ~all frontends use the `pep517` machinery to call the hooks, it's easier to impose an extra requirement on callers, which can largely be changed in one implementation, to allow a bit of extra flexibility for all backends. It's not an entirely ridiculous argument. But frontends aren't required to use this library, and I don't give much weight to 'there will always be more parties on *this* side of the standardised interface' argument, because I've seen how that prediction can be wrong.

**Comment by webknjaz** [comment]
> You're proposing that the signature be `build_wheel(*, wheel_directory, config_settings=None, metadata_directory=None)` which is a change to the PEP.

Yep. I guess a better question would be if it's reasonable to propose changing the PEP.

@pradyunsg currently, I have something like
```python
import functools
import inspect

from another_backend import build_sdist, build_wheel, ...

def kwargize(f):  # ensures that the wrapper in `decorate` always gets kwargs
    sig = inspect.signature(f)  # <-- lots of extra pain under py2 because of bugs in `functools.wraps()` backports + `signature()` also needs a backport

    @functools.wraps(f)
    def w(*args, **kwargs):
        kwargs = dict(kwargs, **sig.bind(*args, **kwargs).arguments)  # map args to kwargs
        return f(**kwargs)
    return w

def decorate(f):  # a generic decorator that is applied to hooks with different signatures all having `config_settings`
    @functools.wraps(f)
    def w(*args, **kwargs):
        assert not args
        config_settings = kwargs['config_settings']  # signature-independent var extraction
        # do stuff
        return f(**kwargs)
    return w

build_sdist = kwargize(decorate(build_sdist))
build_wheel = kwargize(decorate(build_wheel))
```

**Comment by pfmoore** [comment]
> I guess a better question would be if it's reasonable to propose changing the PEP.

IMO, no. Doing so would impose an annoying requirement on *all* callers just to help a fairly obscure use case. I'd be -1 on such a proposal.

I fail to see why you can't just do a bit more work to introspect the arguments based on the hook that's being called. Your example code is barely readable to me. Compare:

```python
def build_sdist(sdist_directory, config_settings=None):
    # do stuff with config_settings
    another_backend.build_sdist(sdist_directory, config_settings)

def build_wheel(wheel_directory, config_settings=None, metadata_directory=None):
    # do stuff with config_settings
    another_backend.build_wheel(wheel_directory, config_setting, metadata_directory)
```

That seems far clearer, and a lot simpler.

**Comment by webknjaz** [comment]
Mostly because I'm lazy and don't like much copy-paste :man_shrugging:

But I get why you wouldn't want to add more restrictions to the PEP. Good to have such a discussion documented.

P.S. I guess my snippet is complex also because I wanted to fit in a lot of details in one block :)
